api/app/src/routes/medical/organization.ts

```diff
@@ -20,22 +25,72 @@ router.post(
   asyncHandler(async (req: Request, res: Response) => {
     const cxId = getCxIdOrFail(req);
 
-    // TODO: parse this into model
-    const orgData = req.body;
+    const reqOrgData = organizationSchema.parse(req.body);
+
+    // TODO: I FIGURED THESE 3 WOULD BE GOOD TO STORE IN DB
+    // CAN ADD GATEWAYS OR TECHNICAL CONTACT ETC.
+    const localOrgPayload = {
+      name: reqOrgData.name,
+      type: "Hospital",
+      locations: reqOrgData.locations,
+    };
 
     // update if this is an existing org
-    let org: Organization;
-    if (orgData.id) {
-      const data = { ...orgData };
+    let localOrg: Organization;
+
+    if (reqOrgData.id) {
+      const data = { ...reqOrgData };
       delete data.id;
-      org = await updateOrganization({ id: orgData.id, cxId, data });
+      localOrg = await updateOrganization({ id: reqOrgData.id, cxId, data: localOrgPayload });
     } else {
-      org = await createOrganization({ cxId, data: orgData });
+      localOrg = await createOrganization({ cxId, data: localOrgPayload });
     }
 
-    // TODO: create or update organization in CW as well
+    const cwOrgPayload: CWOrganization = {
+      ...localOrgPayload,
+      // NOTE: IN STAGING IF THE ID ALREADY EXISTS IT WILL SAY INVALID ORG WHEN CREATING
+      organizationId: `${CW_ID_PREFIX}${localOrg.id}`,
+      homeCommunityId: `${CW_ID_PREFIX}${localOrg.id}`,
+      patientIdAssignAuthority: `${CW_ID_PREFIX}${localOrg.id}`,
+      displayName: reqOrgData.name,
+      memberName: "Metriport",
+      securityTokenKeyType: "BearerKey",
+      isActive: true,
+      gateways: [
+        {
+          serviceType: "XCA_Query",
+          gatewayType: "R4",
+          endpointLocation: Config.getGatewayEndpointLocation(),
+        },
+      ],
+      authorizationInformation: {
+        authorizationServerEndpoint: Config.getGatewayAuthorizationServerEndpoint(),
+        clientId: Config.getGatewayAuthorizationClientId(),
+        clientSecret: Config.getGatewayAuthorizationClientSecret(),
+        documentReferenceScope: "fhir/document",
+        binaryScope: "fhir/document",
+      },
+      technicalContacts: [
+        {
+          name: "Metriport Team",
+          title: "Engineering",
+          email: "support@metriport.com",
+          phone: "(415)-941-3282",
+        },
+      ],
+    };
+
+    try {
+      if (reqOrgData.id) {
+        await commonWellMember.updateOrg(queryMeta, cwOrgPayload, `${reqOrgData.id}`);
+      } else {
+        await commonWellMember.createOrg(queryMeta, cwOrgPayload);
+      }
```

Comment:
Should we do this before updating our DB, and if it fails, not make any mods to our DB?

Reply:
This is super tricky, indeed:
- change local and what if it fails when updating remote?
- change remote and what if it fails updating local?

This is a classic problem w/ distributed systems.

Ideally, we would have a way to do it local first and, if the remote update fails, we would have a way to tell its "dirty" or mismatching. Could have a flag to indicate its sync'ed, could make the record locked until unblocked by Ops (who used a tool to sync local and remote), etc.

Since there are more variables with the external system, I'd first update local and then remote.

Regardless of the situation, we need a way to be warned about this and a tool to fix those situations.

Reply:
Actually, what I was about to suggest on Patient (and obviously we should align on this), is we do the CW synchronization in an asynchronous way, so our users (UI and API) don't have to wait for the response of CW.

Some operations with CW take more than a second, and when we account for the fact this is one of the partners we want to integrate with, it becomes clear we should separate our logic from theirs.

This is related to [a discussion we had before](https://metriport.slack.com/archives/C04DMKE9DME/p1677177674398529?thread_ts=1677174290.564819&cid=C04DMKE9DME) about having our own copy of data or not (we were referring to links). In theory, we could make synchronous requests to CW to retrieve data when needed, and asynchronous when updating data.

But good if we prefer to keep all synchronous for now and upgrade to an asynchronous approach later ([this](https://github.com/metriport/metriport-internal/issues/393) is also related, event based extensions to our core logic, supporting integrations without contaminating our domain).

Reply:
[taken to slack](https://metriport.slack.com/archives/C04DMKE9DME/p1677686900952009)